ouroboros-consensus-cardano/tools/db-validator/Main.hs

```diff
@@ -0,0 +1,263 @@
+  , configFile           :: w ::: [FilePath]      <?> "Configuration file (e.g. mainnet-genesis.json)"
+  , systemStart          :: w ::: Maybe UTCTime   <?> "System start time"
+  , requiresNetworkMagic :: w ::: Bool            <?> "Expecto patronum?"
+  , genesisHash          :: w ::: Hash CB.Raw     <?> "Expected genesis hash"
+  , blockType            :: w ::: Maybe BlockType <?> "type of network"
+  , verbose              :: w ::: Bool            <?> "Enable verbose logging"
+  , onlyImmDB            :: w ::: Bool            <?> "Validate only the immutable DB (e.g. do not do ledger validation)"
+  } deriving (Generic)
+
+instance ParseRecord (Args Wrapped)
+
+deriving instance Show (Args Unwrapped)
+
+data BlockType = Byron | Shelley | Cardano
+  deriving (Generic, Show, Read)
+
+main :: IO ()
+main = do
+  (cmd :: Args Unwrapped) <- unwrapRecord "DB validator"
+  let Args{ dbDir, blockType } = cmd
+  case blockType of
+    Just Byron   -> validate cmd (Proxy @ByronBlock)
+    Just Shelley -> validate cmd (Proxy @(ShelleyBlock TPraosStandardCrypto))
+    Just Cardano -> validate cmd (Proxy @(CardanoBlock TPraosStandardCrypto))
+    Nothing -> do
+      -- check the dbmarker of the db if the block type is not specified.
+      protocolMagicId <- readDBMarker dbDir
+      case unProtocolMagicId protocolMagicId of
+        764824073  -> validate cmd (Proxy @ByronBlock)
+        1097911063 -> validate cmd (Proxy @ByronBlock)
+        42         -> validate cmd (Proxy @(ShelleyBlock TPraosStandardCrypto))
+        _          -> error $ "unsupported protocolMagicId: " ++ show protocolMagicId
+
+readDBMarker :: FilePath -> IO ProtocolMagicId
+readDBMarker dbPath = do
+    bs <- BS.readFile markerPath
+    protocolMagicId <- runExceptT $ dbMarkerParse markerPath bs
+    either
+      (\err -> error $
+        "failed to parse protocolMagicId from db Marker file. Error " ++ show err)
+      return
+      protocolMagicId
+  where
+    markerPath :: String
+    markerPath = dbPath FilePath.</> Text.unpack dbMarkerFile
+
+validate :: forall blk. (HasProtocolInfo blk, Node.RunNode blk, Show (Header blk))
+         => Args Unwrapped -> Proxy blk ->  IO ()
+validate Args {..} _ = do
+    protocolInfo <- mkProtocolInfo @blk configFile genesisHash requiresNetworkMagic
+    validateChainDb dbDir protocolInfo onlyImmDB verbose
+
+validateChainDb
+  :: (Node.RunNode blk, Show (Header blk))
+  => FilePath -- ^ DB directory
+  -> ProtocolInfo IO blk
+  -> Bool -- Immutable DB only?
+  -> Bool -- Verbose
+  -> IO ()
+validateChainDb dbDir protocolInfo onlyImmDB verbose =
+    withRegistry $ \registry -> do
+      let chainDbArgs = mkChainDbArgs registry InFuture.dontCheck
+          (immDbArgs, _, _, _) = fromChainDbArgs chainDbArgs
+      if onlyImmDB then
+        ImmDB.withImmDB immDbArgs $ \immDB -> do
+          immDbTipPoint <- ImmDB.getPointAtTip immDB
+          putStrLn $ "DB tip: " ++ show immDbTipPoint
```

Comment:
Note this used to be `condense`, but I changed to `show` because we don't have all Condense instances needed.

Reply:
That's fine